### ISSUE TITLE
`UnixStream::connect` regression after mio v1 upgrade

### ISSUE THREAD

**Issue by littledivy** [tokio-rs/tokio#6767]
`UnixStream::connect` with abstract address returns an Err starting from mio 1.0.0. 

```
called `Result::unwrap()` on an `Err` value: Error { kind: InvalidInput, message: "paths must not contain interior null bytes" }
```

```rust
// tokio 1.39.2

use tokio::net::UnixStream;

#[tokio::main]
async fn main() {
  UnixStream::connect("\0aaa").await.unwrap();
}
```

Reproduction with mio:

```rust
// mio = { version = "=1.0.0", features = ["net"] } // error
// mio = { version = "=0.8.11", features = ["net", "os-poll"] } // works

use mio::net::UnixStream;

fn main() {
  UnixStream::connect("\0aaa").unwrap();
}
```

Expected behavior: it connects

**Comment by cailloumajor** [comment]
I have the same issue with `tokio::net::UnixListener::bind`. The standard library has `std::os::unix::net::UnixListener::bind_addr` and `std::os::unix::net::SocketAddr::from_abstract_name` to allow using an abstract socket name.

**Comment by mox692** [comment]
Yes, we should probably align with [mio's change](https://github.com/tokio-rs/mio/pull/1749).

**Comment by Darksonn** [comment]
Damn, I even looked into this and concluded that we didn't need to do anything. I guess I was wrong.

I'm not convinced we can get rid of our `SocketAddr` type without it being breaking.

**Comment by mox692** [comment]
Can we fallback to std's `from_abstract_name` if the beginning of the path is `\0`? (maybe it's not a very smart way though)

```
let addr = if path.as_ref().starts_with(b"\0") {
	StdSocketAddr::from_abstract_name(path.as_ref().as_bytes()).unwrap()
} else {
	StdSocketAddr::from_pathname(path).unwrap()
};
let stream = mio::net::UnixStream::connect_addr(&addr)?;
```

In the longer term, we may want to depricate our own `SocketAddr`.

**Comment by Darksonn** [comment]
Let's do that.

**Comment by mox692** [comment]
This was resolved in https://github.com/tokio-rs/tokio/pull/6772 and shipped as [v1.39.3](https://github.com/tokio-rs/tokio/releases/tag/tokio-1.39.3)
